### PR TITLE
bump contracts and conform to stricter transaction formats

### LINKED ISSUE #1176: Strict RLP decoding / Enforce PaymentTransaction according to spec
### Description
Recent efforts to achieve a 1:1 mapping from an RLP tx to a decoded PaymentTransaction have recently been merged into the root chain contracts ( PR https://github.com/omisego/plasma-contracts/pull/476). This issue proposes to implement the same mechanisms to achieve a 1:1 mapping and to decode/encode only well-formed PaymentTransactions. 

### Enforce non overlapping RLP decoding
Achieving a 1:1 relationship between decoded and encoded RLP message is not trivial and there are some edge cases that can easily be overlooked. Check out the test inputs in https://github.com/omisego/plasma-contracts/blob/182908d4c7ef669e426356c0bce528a5ebe75593/plasma_framework/test/src/utils/fixture/ and the RLP related test runners in https://github.com/omisego/plasma-contracts/blob/182908d4c7ef669e426356c0bce528a5ebe75593/plasma_framework/test/src/utils/. The implementation in Elixir should cover all of the tests that have been specific on the plasma-contracts.

#### Define Data types 

The current RLPReader implements RLP conform decoding as specified here https://github.com/ethereum/wiki/wiki/RLP.  It also introduces 3 data types that are application specific. 

- [x] RLPReader.toUint():  treats leading zeros as invalid. Must be between 1 and 32 bytes. 
- [x] RLPReader.toAddress() requires the item data to be exactly 20 bytes.
- [x] RLPReader.toBytes32() requires the item data to be exactly 32 bytes.

#### Decode/Encode only valid PaymentTransaction 

During PaymentTransaction decoding it is also checked that the tx is well formed and according to spec: 

- [x]  Transaction inputs must be decoded as bytes32
- [x]  Transaction inputs cannot be empty (all zeros)
- [x]  Transaction outputs cannot be empty (a list that has zero items)
- [x]  Transaction output.amount must be greater than zero
- [x]  Transaction output.outputType must be greater than zero
- [x]  Transaction metadata must be bytes32

CC @kevsul feel free to add details.

### PR DESCRIPTION
**Pull request by pdobacz**
- fixes #1176 
 - supersedes #1160 in adopting the Exit multigetter functionality and as such, works towards resolving https://github.com/omisego/security-issues/issues/7, IFE support still missing

## Overview

As title and see issue #1176 for a detailed rundown of what "stricter" means.

## Changes

- contracts bumped (**TODO**: do the plasma-deployer victory dance)
- stricter encodings done, commit by commit, following #1176 
- when moving forward, I was also running #1164 conformance tests to see if anything slips by untested
- watcher now conforms to the multi-getter of exit structures from the contract by @InoMurko. However, this required https://github.com/poanetwork/ex_abi/issues/22 to work properly. To not stall this, a workaround for the failing decoding was put in `root_chain.ex` (separate commit :point_down: )
- unfortunately, some tests liked to test using no-outputs transactions, those had to be trivially reworked. No tests were changed significantly or removed (unless they tested now illegal behaviors)

## Testing

as usual + we'll need to deploy new contracts and e2e test, when that's ready.

### CODE REVIEW

File: apps/omg/test/omg/state/transaction_test.exs [hunk 1]
```diff
@@ -201,7 +199,10 @@ defmodule OMG.State.TransactionTest do
                Transaction.Recovered.recover_from(ExRLP.encode([sigs, @payment_marker, inputs, outputs, ""]))
 
       assert {:error, :malformed_metadata} =
-               Transaction.Recovered.recover_from(ExRLP.encode([sigs, @payment_marker, inputs, outputs, <<0::288>>]))
+               Transaction.Recovered.recover_from(ExRLP.encode([sigs, @payment_marker, inputs, outputs, <<1::224>>]))
+
+      assert {:error, :malformed_metadata} =
+               Transaction.Recovered.recover_from(ExRLP.encode([sigs, @payment_marker, inputs, outputs, <<2::288>>]))
```
Comment (InoMurko) [inline apps/omg/test/omg/state/transaction_test.exs]: this is testing module `Transaction.Recovered`, why is this in `transaction_test.exs`?

File: apps/omg_eth/lib/omg_eth/root_chain.ex [hunk 2]
```diff
@@ -70,21 +70,53 @@ defmodule OMG.Eth.RootChain do
...
+      contract,
+      "standardExits(uint160[])",
+      [exit_ids],
+      static_array_return_types
+    )
+  end
+
+  # TODO: see above in where it is called - temporary function
+  defp call_contract_manual_standard_exits(contract, signature, args, return_types) do
+    data = signature |> ABI.encode(args)
```
Comment (InoMurko) [inline apps/omg_eth/lib/omg_eth/root_chain.ex]:
```suggestion
    data = ABI.encode(signature, args)
```

File: apps/omg_watcher/test/omg_watcher/exit_processor/piggyback_test.exs [hunk 3]
```diff
@@ -162,36 +163,54 @@ defmodule OMG.Watcher.ExitProcessor.PiggybackTest do
...
 
       assert {:ok,
               [
                 %Event.PiggybackAvailable{
                   available_inputs: [%{index: 1}],
+                  available_outputs: [%{index: 0}]
+                }
+              ]} =
+               %ExitProcessor.Request{blknum_now: 1000, eth_height_now: 5}
+               |> check_validity_filtered(processor, only: [Event.PiggybackAvailable])
```
Comment (InoMurko) [inline apps/omg_watcher/test/omg_watcher/exit_processor/piggyback_test.exs]:
```suggestion
              check_validity_filtered(%ExitProcessor.Request{blknum_now: 1000, eth_height_now: 5}, processor, only: [Event.PiggybackAvailable])
```

File: apps/omg_watcher/test/omg_watcher/exit_processor/piggyback_test.exs [hunk 4]
```diff
@@ -314,7 +333,7 @@ defmodule OMG.Watcher.ExitProcessor.PiggybackTest do
       tx_blknum = 3000
 
       # 2. transaction which spends that piggybacked output
-      comp = TestHelper.create_recovered([{tx_blknum, 0, 0, alice}], [])
+      comp = TestHelper.create_recovered([{tx_blknum, 0, 0, alice}], [{alice, @eth, 1}])
```
Comment (InoMurko) [inline apps/omg_watcher/test/omg_watcher/exit_processor/piggyback_test.exs]: why did we have to start adding outputs?

Comment (pdobacz) [reply apps/omg_watcher/test/omg_watcher/exit_processor/piggyback_test.exs]: see description, no-output txs are illegal now

File: apps/omg/test/omg/state/transaction_test.exs (hunk 1 above)
Comment (pdobacz) [reply apps/omg/test/omg/state/transaction_test.exs]: yeah, leftovers. Did a quick split

File: apps/omg_watcher/test/omg_watcher/exit_processor/core/state_interaction_test.exs [hunk 5]
```diff
@@ -128,7 +128,7 @@ defmodule OMG.Watcher.ExitProcessor.Core.StateInteractionTest do
 
   test "only asking for spends concerning ifes",
        %{alice: alice, processor_empty: processor, state_empty: state_empty} do
-    processor = processor |> start_ife_from(TestHelper.create_recovered([{1, 0, 0, alice}], []))
+    processor = processor |> start_ife_from(TestHelper.create_recovered([{1, 0, 0, alice}], [{alice, @eth, 1}]))
```
Comment (InoMurko) [inline apps/omg_watcher/test/omg_watcher/exit_processor/core/state_interaction_test.exs]: as per Elixir style guide, single pipes are discouraged
